### ISSUE TITLE
Minimization issue with c-vars as head

### ISSUE THREAD

**Issue by mudbri** [ravel-net/pyotr#13]
Minimization algorithm is not working properly when c-variables are in the head. Consider the code below
```
p1 = "l(a,b) :- l(y,c), k(d,y,e), l(f,g)\nl(3,4) :- l(1,3), k(2,1,3), l(1,5)"
p2 = "l(3,4) :- l(1,3), k(2,1,3), l(1,5)"

program1 = DT_Program(p1, {"l":["int4_faure", "int4_faure"], "k":["int4_faure", "int4_faure", "int4_faure"]}, domains={'a':[1,2,3,4,5,6,7],'b':[1,2,3,4,5,6,7], 'c':[1,2,3,4,5,6,7],'d':[1,2,3,4,5,6,7],'e':[1,2,3,4,5,6,7],'f':[1,2,3,4,5,6,7],'g':[1,2,3,4,5,6,7],'y':[1,2,3,4,5,6,7]}, c_variables=['a','b','c','d','e','f','g','y'], reasoning_engine='z3', reasoning_type='Int', datatype='int4_faure', simplification_on=True, c_tables=["l","k"])

program1.minimize()
print(program1)
```
The result of minimization is l(3,4) :- l(1,5), which is clearly wrong

**Comment by mudbri** [comment]
This issue is most probably triggered by incorrect treatment of atoms with c-variables when checking containment. Another related issue can be seen in the following example:
```
p1 = "l(3,4) :- l(1,3), k(2,1,3), l(1,5)"
p2 = "l(3,4) :- l(y,c), k(d,y,e), l(f,g)"
```
Here, if variables in p2 are treated as normal variables, then p1 does not contain p2. However, when they are treated as c-variables, then our implementation shows that p1 does contain p2. This is because the c-variables in p2 will be mapped to themselves with no condition. Since all constants are equal to c-variable, the target head will be generated. The problem is that we do not see if equating c-variable to a constant is implied by the condition associated with the atom, which is what we should do. Consider another program p3:
`p3 = "l(3,4) :- l(y,c)[y == 1 And c == 3], k(d,y,e) [d == 2 And y == 1 And e == 3], l(f,g) [f == 1 and g == 5]"`
Here, p1 does contain p3. The only difference is that any assignment of c-variable to constant in p1 is implied by the condition associated with that c-variable in p3.

For clarity, consider another program p4:
`p4 = "l(3,4) :- l(y,c)[y == 1 And c >= 3], k(d,y,e) [d == 2 And y == 1 And e == 3], l(f,g) [f == 1 and g == 5]"`
Here, p1 does not contain p4, since the condition `c >= 3` does not imply the equality `c == 3`

Note: A potential solution to [issue](https://github.com/ravel-net/pyotr/issues/16) here could be to treat variables as c-variables with no condition and using the implication method

**Comment by mudbri** [comment]
Fixed in commit 4b4628af17fe2e7b695900fd63e18e913330de25

Note that in this commit, we check for equivalence instead of implication between the conditions in the tuple and the condition in the head. This change was made in light of the example in the previous comment.

**Comment by mudbri** [comment]
1. Use the term substitution instead of mapping
2. We can generalize variables into just c-variables with no condition
3. Note that variables cannot be substituted by any other constant (which is why we initiate them to distinct constants)
